src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarHeader.Read.cs

```diff
@@ -28,45 +30,162 @@ internal bool TryGetNextHeader(Stream archiveStream, bool copyData)
 
             archiveStream.ReadExactly(buffer);
 
-            try
+            // Confirms if v7 or pax, or tentatively selects ustar
+            if (!TryReadCommonAttributes(buffer))
             {
-                // Confirms if v7 or pax, or tentatively selects ustar
-                if (!TryReadCommonAttributes(buffer))
+                return false;
+            }
+
+            // Confirms if gnu, or tentatively selects ustar
+            ReadMagicAttribute(buffer);
+
+            if (_format != TarEntryFormat.V7)
+            {
+                // Confirms if gnu
+                ReadVersionAttribute(buffer);
+
+                // Fields that ustar, pax and gnu share identically
+                ReadPosixAndGnuSharedAttributes(buffer);
+
+                Debug.Assert(_format is TarEntryFormat.Ustar or TarEntryFormat.Pax or TarEntryFormat.Gnu);
+                if (_format == TarEntryFormat.Ustar)
+                {
+                    ReadUstarAttributes(buffer);
+                }
+                else if (_format == TarEntryFormat.Gnu)
                 {
-                    return false;
+                    ReadGnuAttributes(buffer);
                 }
+                // In PAX, there is nothing to read in this section (empty space)
+            }
+
+            ProcessDataBlock(archiveStream, copyData);
+
+            ArrayPool<byte>.Shared.Return(rented);
+            return true;
+        }
+
+        // Asynchronously attempts read all the fields of the next header.
+        // Throws if end of stream is reached or if any data type conversion fails.
+        // Returns true if all the attributes were read successfully, false otherwise.
+        internal static async ValueTask<(bool, TarHeader)> TryGetNextHeaderAsync(Stream archiveStream, bool copyData, TarEntryFormat initialFormat, CancellationToken cancellationToken)
```

Comment:
I currently don't have a good suggestion for how this method could be refactored to reduce code duplication. I'll try to play with the code once your PR gets merged.

Reply:
I know. TarHeader is the problem,  since it is a struct. I may have to refactor the code later to get rid of TarHeader completely,  move everything to TarEntry.